res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+    if (value) {
+        value = 0x7F;
+    } else {
+        value = 0x00;
+    }
+    var channel = group.match(/^\[Channel(\d+)\]$/)[1];
+    midi.sendShortMsg(0x9F, 0x00+(channel-1), value);
+};
+
+PioneerDDJ400.toggleLight = function(midiIn, active) {
+    midi.sendShortMsg(midiIn.status, midiIn.data1, active ? 0x7F : 0);
+};
+
+//
+// Init
+//
+
+PioneerDDJ400.init = function() {
+    // init controller
+
+    // show focus buttons on Effect Ract 1 only
+    engine.setValue("[EffectRack1_EffectUnit1]", "show_focus", 1);
+
+    // Connect the VU-Meter LEDS
+    engine.makeConnection("[Channel1]", "VuMeter", PioneerDDJ400.vuMeterUpdate);
+    engine.makeConnection("[Channel2]", "VuMeter", PioneerDDJ400.vuMeterUpdate);
+
+    // reset vumeter
+    PioneerDDJ400.toggleLight(PioneerDDJ400.lights.deck1.vuMeter, false);
+    PioneerDDJ400.toggleLight(PioneerDDJ400.lights.deck2.vuMeter, false);
+
+    // enable soft takeover for rate controls and FX level/depth
+    engine.softTakeover("[Channel1]", "rate", true);
+    engine.softTakeover("[Channel2]", "rate", true);
+    engine.softTakeover("[EffectRack1_EffectUnit1_Effect1]", "meta", true);
+    engine.softTakeover("[EffectRack1_EffectUnit1_Effect2]", "meta", true);
+    engine.softTakeover("[EffectRack1_EffectUnit1_Effect3]", "meta", true);
+    engine.softTakeover("[EffectRack1_EffectUnit1]", "mix", true);
+
+    // Sampler callbacks
+    for (var i = 1; i <= 16; ++i) {
+        engine.makeConnection("[Sampler" + i + "]", "play", PioneerDDJ400.samplerPlayOutputCallbackFunction);
+    }
+
+    // trigger "track loaded" animations when a track is loaded
+    engine.makeConnection("[Channel1]", "track_loaded", PioneerDDJ400.trackLoadedLED);
+    engine.makeConnection("[Channel2]", "track_loaded", PioneerDDJ400.trackLoadedLED);
+
+    // eye candy : play the "track loaded" animation on both decks at startup
+    midi.sendShortMsg(0x9F, 0x00, 0x7F);
+    midi.sendShortMsg(0x9F, 0x01, 0x7F);
+
+    // resets pad mode to hotcue
+    PioneerDDJ400.performancePads.reset();
+
+    // turn on loop in and out lights
+    PioneerDDJ400.setLoopButtonLights(0x90, 0x7F);
+    PioneerDDJ400.setLoopButtonLights(0x91, 0x7F);
+
+    // handle loop toggle events
+    engine.makeConnection("[Channel1]", "loop_enabled", PioneerDDJ400.loopToggle);
+    engine.makeConnection("[Channel2]", "loop_enabled", PioneerDDJ400.loopToggle);
+
+    // poll the controller for current control positions on startup
+    midi.sendSysexMsg([0xF0, 0x00, 0x40, 0x05, 0x00, 0x00, 0x02, 0x06, 0x00, 0x03, 0x01, 0xf7], 12);
+};
+
+//
+// Channel level lights
+//
+
+PioneerDDJ400.vuMeterUpdate = function(value, group) {
+    var newVal = value * 150;
+
+    switch (group) {
+    case "[Channel1]":
+        midi.sendShortMsg(0xB0, 0x02, newVal);
+        break;
+
+    case "[Channel2]":
+        midi.sendShortMsg(0xB1, 0x02, newVal);
+        break;
+    }
+};
+
+//
+// Tempo sliders
+//
```

Comment:
remove

Reply:
Could we leave these "marker" comments  in please? I find them useful as they stand out when scrolling quickly through the script and add a sense of structure to it.

Reply:
If you really want, sure. I don't think it's very helpful to have comments that pretty much repeat what the code right next to them says.

Reply:
I definitely agree with you in general, and have removed most of the useless comments.

Just that when files get monolithic, I find that clear sections add a sense of structure which helps me maintain things better.

Reply:
Okay, if you find those helpful, then keep them in.